uniffi_bindgen/src/library_mode.rs

```diff
@@ -138,39 +124,68 @@ fn find_components(
     metadata_groups
         .into_values()
         .map(|group| {
-            let package = find_package_by_crate_name(cargo_metadata, &group.namespace.crate_name)?;
-            let mut ci = ComponentInterface::new(&group.namespace.crate_name);
-            if let Some(metadata) = udl_items.remove(&group.namespace.crate_name) {
+            let crate_name = &group.namespace.crate_name;
+            let mut ci = ComponentInterface::new(crate_name);
+            if let Some(metadata) = udl_items.remove(crate_name) {
                 ci.add_metadata(metadata)?;
             };
             ci.add_metadata(group)?;
-            Ok((ci, package))
+            Ok(ci)
         })
         .collect()
 }
 
-fn find_package_by_crate_name(
-    metadata: &cargo_metadata::Metadata,
-    crate_name: &str,
-) -> Result<Package> {
-    let matching: Vec<&Package> = metadata
-        .packages
-        .iter()
-        .filter(|p| {
-            p.targets
-                .iter()
-                .any(|t| t.name.replace('-', "_") == crate_name)
-        })
-        .collect();
-    match matching.len() {
-        1 => Ok(matching[0].clone()),
-        n => bail!("cargo metadata returned {n} packages for crate name {crate_name}"),
+#[derive(Debug, Clone, Default)]
+struct CratePathMap {
+    // path on disk where we will find the crate's source.
+    paths: HashMap<String, Utf8PathBuf>,
+}
+
+impl CratePathMap {
+    // TODO: we probably need external overrides passed in (maybe via
+    // `config_file_override` - what are the semantics of that?)
+    // Anyway: for now we just do this.
+    #[cfg(feature = "cargo_metadata")]
+    fn new() -> Result<Self> {
+        Ok(Self::from(
+            cargo_metadata::MetadataCommand::new()
+                .exec()
+                .context("error running cargo metadata")?,
+        ))
+    }
+
+    #[cfg(not(feature = "cargo_metadata"))]
+    fn new() -> Result<Self> {
+        Ok(Self::default())
+    }
+
+    fn get_path(&self, crate_name: &str) -> Option<&Utf8Path> {
+        self.paths.get(crate_name).map(|p| p.as_path())
+    }
+}
+
+// all this to get a few paths!?
+#[cfg(feature = "cargo_metadata")]
+impl From<cargo_metadata::Metadata> for CratePathMap {
+    fn from(metadata: cargo_metadata::Metadata) -> Self {
+        let paths: HashMap<String, Utf8PathBuf> = metadata
+            .packages
+            .iter()
+            .flat_map(|p| {
+                p.targets.iter().filter(|t| t.is_lib()).filter_map(|t| {
```

Comment:
This seemingly harmless filter does not seem correct. I just investigated why my code generation broke with uniffi 0.28.1 and the answer is that the crate just does not get added to the map due to this anymore.

It has `crate-type = ["staticlib", "cdylib"]` and the `cargo_metadata` crate might not do what you expect (as its documentation suggests differently): `kind` and `crate_types` both are `["staticlib", "cdylib"]` and `is_lib` only returns true if `"lib"` is contained there. From what the `cargo_metadata` interface suggests, the target is of no type (none of the provided `is_*` methods returns true and `kind` does not have any of the values mentioned in its documentation).